packages/core/saved-objects/core-saved-objects-api-server-internal/src/lib/repository.ts

```diff
@@ -762,6 +776,289 @@ export class SavedObjectsRepository implements ISavedObjectsRepository {
+    const { expectedBulkGetResults, namespace } = params;
+    const bulkGetMultiNamespaceDocs = expectedBulkGetResults
+      .filter(isRight)
+      .filter(({ value }) => value.esRequestIndex !== undefined)
+      .map(({ value: { type, id } }) => ({
+        _id: this._serializer.generateRawId(namespace, type, id),
+        _index: this.getIndexForType(type),
+        _source: ['type', 'namespaces'],
+      }));
+
+    const bulkGetMultiNamespaceDocsResponse = bulkGetMultiNamespaceDocs.length
+      ? await this.client.mget(
+          { body: { docs: bulkGetMultiNamespaceDocs } },
+          { ignore: [404], meta: true }
+        )
+      : undefined;
+    // fail fast if we can't verify a 404 response is from Elasticsearch
+    if (
+      bulkGetMultiNamespaceDocsResponse &&
+      isNotFoundFromUnsupportedServer({
+        statusCode: bulkGetMultiNamespaceDocsResponse.statusCode,
+        headers: bulkGetMultiNamespaceDocsResponse.headers,
+      })
+    ) {
+      throw SavedObjectsErrorHelpers.createGenericNotFoundEsUnavailableError();
+    }
+    return bulkGetMultiNamespaceDocsResponse;
+  }
+
+  /**
+   * @returns array of objects sorted by expected delete success or failure result
+   * @internal
+   */
+  private getExpectedBulkDeleteMultiNamespaceDocsResults(
+    params: ExpectedBulkDeleteMultiNamespaceDocsParams
+  ): ExpectedBulkDeleteResult[] {
+    const { expectedBulkGetResults, multiNamespaceDocsResponse, namespace, force } = params;
+    let indexCounter = 0;
+    const expectedBulkDeleteMultiNamespaceDocsResults =
+      expectedBulkGetResults.map<ExpectedBulkDeleteResult>((expectedBulkGetResult) => {
+        if (isLeft(expectedBulkGetResult)) {
+          return { ...expectedBulkGetResult };
+        }
+        const { esRequestIndex: esBulkGetRequestIndex, id, type } = expectedBulkGetResult.value;
+
+        let namespaces;
+
+        if (esBulkGetRequestIndex !== undefined) {
+          const indexFound = multiNamespaceDocsResponse?.statusCode !== 404;
+
+          const actualResult = indexFound
+            ? multiNamespaceDocsResponse?.body.docs[esBulkGetRequestIndex]
+            : undefined;
+
+          const docFound = indexFound && isMgetDoc(actualResult) && actualResult.found;
+
+          // return an error if the doc isn't found at all or the doc doesn't exist in the namespaces
+          if (!docFound) {
+            return {
+              tag: 'Left',
+              value: {
+                id,
+                type,
+                error: errorContent(SavedObjectsErrorHelpers.createGenericNotFoundError(type, id)),
+              },
+            };
+          }
+          // the following check should be redundant since we're retrieving the docs from elasticsearch but we check just to make sure
+          // @ts-expect-error MultiGetHit is incorrectly missing _id, _source
+          if (docFound && !this.rawDocExistsInNamespace(actualResult, namespace)) {
```

Comment:
NIT: we exited previously on the `!docFound` condition, so `docFound` can be removed from this condition